### PR TITLE
doc: Defining migration guide for Independent Shard Scaling feature

### PR DESCRIPTION
**Pull request by AgustinBettati**
## Description

Link to any related issue(s): CLOUDP-260670

**Note**: Advise copying the content into doc preview website for better reading and also verifying formatting: https://registry.terraform.io/tools/doc-preview

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [x] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+      provider_name = "AWS"
+      priority      = 7
+      region_name   = "EU_WEST_1"
+    }
+  }
+}
+```
+
+It is now possible to define a different `instance_size`, and `disk_iops` in the case of AWS, for each individual shard. In the following update we will define an upgraded instance size of M40 only for the first shard. One criteria for defining which shard to scaling independently is by looking into the metrics dashboard in the UI (e.g. https://cloud.mongodb.com/v2/<PROJECT-ID>#/clusters/detail/SymmetricShardedCluster)
+
```
Comment (AgustinBettati) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: currently revising limitations worth mentioning when using ISS.

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+      }
+      provider_name = "AWS"
+      priority      = 7
+      region_name   = "EU_WEST_1"
+    }
+  }
+}
+```
+
+This  updated configuration will trigger a terraform update plan, however the underlying cluster will not face any changes after the apply as both configurations represent a sharded cluster composed of 2 shards.
```
Comment (EspenAlbert) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: extra whitespace after `This`
Maybe change order:
`terraform update plan` -> `terraform plan update`?

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    priority      = 7
+    region_name   = "EU_WEST_1"
+    }
+  }
+}
+```
+
+
+
+This  updated configuration will trigger a terraform update plan, however the underlying cluster will not face any changes after the apply as both configurations represent a geo sharded cluster with 2 zones and 2 shards in each one.
```
Comment (EspenAlbert) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: same as above

**Review by EspenAlbert — APPROVED** [review]
Great job!

File: website/docs/guides/1.18.0-upgrade-guide.html.markdown [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+- Deprecations in `mongodbatlas_advanced_cluster` resource and data sources:
+    - `replication_specs.*.num_shards`: The `replication_specs` list now supports defining an object for each inidividual shard and this new schema is favoured over using `num_shards` attribute. For more details and migration guidelines please reference [advanced_cluster - Migration to new sharding schema and leveraging Independent Shard Scaling](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema).
+    - `disk_size_gb`: The same attribute is now defined under `replication_specs.*.region_configs.*.(electable_specs|analytics_specs|read_only_specs).disk_size_gb`. Replacing the root value into existing inner specs will have no change in the underlying cluster. The motivation behind this change in location is to align with the new API schema and facilitate new features related to independent storage size scaling in the future.
+    - `replication_specs.*.id`: This attribute was being used by `cloud_backup_schedule` to identify cluster zones. As of 1.18.0 `cloud_backup_schedule` can reference cluster zones using the new `zone_id` attribute.
```
Comment (maastha) [inline website/docs/guides/1.18.0-upgrade-guide.html.markdown]:
```suggestion
    - `replication_specs.*.id`: This attribute was being used by `mongodbatlas_cloud_backup_schedule` resource to identify cluster zones. As of 1.18.0 `mongodbatlas_cloud_backup_schedule` resource can reference cluster zones using the new `zone_id` attribute.
```

File: website/docs/guides/1.18.0-upgrade-guide.html.markdown [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    - `advanced_configuration.default_read_concern`: This attribute only has one valid value.
```
Comment (maastha) [inline website/docs/guides/1.18.0-upgrade-guide.html.markdown]:
```suggestion
    - `advanced_configuration.default_read_concern`: MongoDB 5.0 and later clusters default to `local`. To use a custom read concern level, please refer to your driver documentation.
```

Comment (AgustinBettati) [reply website/docs/guides/1.18.0-upgrade-guide.html.markdown]: thank you for the edit and reaching out to Jack for more details!

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown (hunk 1 above)
Comment (AgustinBettati) [reply website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: verified it is supported for GCP and Azure as well, and disk_iops is only supported for AWS.

File: website/docs/guides/1.18.0-upgrade-guide.html.markdown [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+---
+
+# MongoDB Atlas Provider 1.18.0: Upgrade and Information Guide
+
+The Terraform MongoDB Atlas Provider version 1.18.0 has a number of new and exciting features.
+
+**New Resources, Data Sources, and Features:**
+
+- Sharded and geo-sharded clusters defined with `mongodbatlas_advanced_cluster` are now capable of scaling the instance size and disk IOPS independently for each individual shard. For more details and migration guidelines please reference [advanced_cluster - Migration to new sharding schema and leveraging Independent Shard Scaling](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema). As part of these changes 2 new attributes have been included:
+    - New attribute `replication_specs.*.zone_id` is added in the resource and data sources to enable identifying the zone of each `replication_specs` object.
```
Comment (jvincent-mongodb) [inline website/docs/guides/1.18.0-upgrade-guide.html.markdown]: Is this zone specific to geo-sharded clusters?

File: website/docs/guides/1.18.0-upgrade-guide.html.markdown [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    - New attribute `use_replication_spec_per_shard` defined in both data sources to configure if the users want to obtain `replication_specs` objects for each shard.
```
Comment (jvincent-mongodb) [inline website/docs/guides/1.18.0-upgrade-guide.html.markdown]: Which data sources?

File: website/docs/guides/1.18.0-upgrade-guide.html.markdown [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+**Deprecations:**
+
+- Deprecations in `mongodbatlas_advanced_cluster` resource and data sources:
+    - `replication_specs.*.num_shards`: The `replication_specs` list now supports defining an object for each inidividual shard and this new schema is favoured over using `num_shards` attribute. For more details and migration guidelines please reference [advanced_cluster - Migration to new sharding schema and leveraging Independent Shard Scaling](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema).
```
Comment (jvincent-mongodb) [inline website/docs/guides/1.18.0-upgrade-guide.html.markdown]: Run-on sentence:  ... each individual shard. This new schema ...

Comment (jvincent-mongodb) [reply website/docs/guides/1.18.0-upgrade-guide.html.markdown]: missing comma: ... migration guidelines, please ...

File: website/docs/guides/1.18.0-upgrade-guide.html.markdown [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    - `disk_size_gb`: The same attribute is now defined under `replication_specs.*.region_configs.*.(electable_specs|analytics_specs|read_only_specs).disk_size_gb`. Replacing the root value into existing inner specs will have no change in the underlying cluster. The motivation behind this change in location is to align with the new API schema and facilitate new features related to independent storage size scaling in the future.
```
Comment (jvincent-mongodb) [inline website/docs/guides/1.18.0-upgrade-guide.html.markdown]: Does this mean that if a cluster is currently deployed with a value of "X" set to `disk_size_gb` and that same value is set to the new field location, the cluster configuration will not change?

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+- [Overview of schema changes](#overview)
+- [Migrating existing advanced_cluster type SHARDED](#migration-sharded)
+- [Migrating existing advanced_cluster type GEOSHARDED](#migration-geosharded)
+- [Migrating existing advanced_cluster type REPLICASET](#migration-replicaset)
+- [Leveraging Independent Shard Scaling](#leveraging-iss)
+
+<a id="overview"></a>
+# Overview of schema changes
+
+`replication_specs` attribute has been modified to now be able to represent each individual shard of a cluster with a unique replication spec element. This implies that when using the new sharding schema the existing attribute `num_shards` will no longer be defined, and instead the number of shards will be defined by the number of `replication_specs` elements.
```
Comment (jvincent-mongodb) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: This implies that ... => As such, ...

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+      }
+      provider_name = "AWS"
+      priority      = 7
+      region_name   = "EU_WEST_1"
+    }
+  }
+}
+```
+
+In order to update our configuration to the new schema we will remove the use of `num_shards` and add a new identical `replication_specs` element for each shard. Note that these 2 changes must be done at the same time.
```
Comment (jvincent-mongodb) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: missing comma: In order to update our configuration to the new schema, we will ...

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown (hunk 2 above)
Comment (jvincent-mongodb) [reply website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: Suggested rephrasing:

This updated configuration will trigger a `terraform update plan`. However, the underlying cluster will not change after the apply, as both configurations represent a sharded cluster composed of two shards.

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+    }
+    provider_name = "AWS"
+    priority      = 7
+    region_name   = "EU_WEST_1"
+    }
+  }
+}
+```
+
+In order to update our configuration to the new schema we will remove the use of `num_shards` and add a new identical `replication_specs` element for each shard. Note that these 2 changes must be done at the same time.
```
Comment (jvincent-mongodb) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: missing comma: to the new schema, we will remove ...

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown (hunk 3 above)
Comment (jvincent-mongodb) [reply website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: Suggested rephrasing:

This updated configuration will trigger a `terraform update plan`. However, the underlying cluster will not change after the apply, as both configurations represent a geo-sharded cluster with two shards in each of the two zones that the cluster spans.

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+            }
+            provider_name = "AZURE"
+            priority      = 7
+            region_name   = "US_EAST"
+        }
+    }
+}
+```
+
+To transition a replica set to sharded cluster 2 separate updates must be applied. First the cluster type must be adjusted to SHARDED, and apply this change without any other additional changes.
```
Comment (jvincent-mongodb) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: It seems like there is only one update being made here?

Comment (jvincent-mongodb) [reply website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: Oh, I see. There is another step below. Maybe rephrase to something like:

First, update the cluster type to `SHARDED`, and apply this change to the cluster.

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+            region_name   = "US_EAST"
+        }
+    }
+}
+```
+
+<a id="leveraging-iss"></a>
+## Leveraging Independent Shard Scaling 
+
+Prerequisite: The new sharding schema must be used, meaning the advanced_cluster configuration is not using `num_shards` and therefore each shard is represented with a unique `replication_specs` element. Please refer to documentation above to transition into this schema if not already.
```
Comment (jvincent-mongodb) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: Suggested rephrasing:

The new sharding schema must be used. Each shard must be represented with a unique `replication_specs` element, as illustrated in the following example.

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+
+Considering the following configuration of a SHARDED cluster that has a symmetric configuration for its 2 shards:
```
Comment (jvincent-mongodb) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: Suggestion: remove this sentence.

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+      }
+      provider_name = "AWS"
+      priority      = 7
+      region_name   = "EU_WEST_1"
+    }
+  }
+}
+```
+
+It is now possible to define a different `instance_size`, and `disk_iops` in the case of AWS, for each individual shard. In the following update we will define an upgraded instance size of M40 only for the first shard. One criteria for defining which shard to scaling independently is by looking into the metrics dashboard in the UI (e.g. https://cloud.mongodb.com/v2/<PROJECT-ID>#/clusters/detail/SymmetricShardedCluster)
```
Comment (jvincent-mongodb) [inline website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: Suggested rephrasing: With each shard's `replication_specs` defined independently, we can now define distinct `instance_size`, and `disk_iops` values for each shard in the cluster. In the following example, we define an upgraded instance size of `M40` only for the first shard in the cluster. 

Consider reviewing the Metrics Dashboard in the MongoDB Atlas UI (e.g. https://cloud.mongodb.com/v2/<PROJECT-ID>#/clusters/detail/SymmetricShardedCluster) for insight into how each shard within your cluster is currently performing, which will inform any shard-specific resource allocation changes you might require.

**Review by jvincent-mongodb — CHANGES_REQUESTED** [review]
I left a few questions and comments for your consideration. Thanks!

File: website/docs/guides/1.18.0-upgrade-guide.html.markdown (hunk 6 above)
Comment (AgustinBettati) [reply website/docs/guides/1.18.0-upgrade-guide.html.markdown]: is will actually be populated for all cluster types (replica set, sharded, geo sharded)

File: website/docs/guides/1.18.0-upgrade-guide.html.markdown (hunk 7 above)
Comment (AgustinBettati) [reply website/docs/guides/1.18.0-upgrade-guide.html.markdown]: adjusted to add explicit mention: defined in `mongodbatlas_advanced_cluster` data sources

File: website/docs/guides/1.18.0-upgrade-guide.html.markdown (hunk 9 above)
Comment (AgustinBettati) [reply website/docs/guides/1.18.0-upgrade-guide.html.markdown]: that is correct

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown (hunk 14 above)
Comment (AgustinBettati) [reply website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: sounds better, adjusted

File: website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown (hunk 16 above)
Comment (AgustinBettati) [reply website/docs/guides/advanced-cluster-new-sharding-schema.html.markdown]: thanks for the suggestion, makes it more clear.

**Review by lantoli — APPROVED** [review]
note that there will be conflicts with master as doc structure has changed folder and file extension

**Review by jvincent-mongodb — APPROVED** [review]
LGTM!
